clients/src/test/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImplTest.java

```diff
@@ -790,6 +791,297 @@ public void testOnSubscriptionUpdatedTransitionsToJoiningOnlyIfNotInGroup() {
+                ConsumerRebalanceListenerMethodName.ON_PARTITIONS_LOST,
+                topicPartitions("topic1", 0)
+        );
+
+        // Step 4: Receive ack and make sure we're done and our listener was called appropriately
+        membershipManager.onHeartbeatRequestSent();
+        assertEquals(MemberState.JOINING, membershipManager.state());
+
+        assertEquals(0, listener.revokedCounter.get());
+        assertEquals(0, listener.assignedCounter.get());
+        assertEquals(1, listener.lostCounter.get());
+    }
+
+    private ConsumerRebalanceListenerInvoker consumerRebalanceListenerInvoker() {
+        ConsumerCoordinatorMetrics coordinatorMetrics = new ConsumerCoordinatorMetrics(
+                subscriptionState,
+                new Metrics(),
+                "test-");
+        return new ConsumerRebalanceListenerInvoker(
+                new LogContext(),
+                subscriptionState,
+                new MockTime(1),
+                coordinatorMetrics
+        );
+    }
+
+    private SortedSet<TopicPartition> topicPartitions(String topicName, int... partitions) {
+        SortedSet<TopicPartition> topicPartitions = new TreeSet<>(new Utils.TopicPartitionComparator());
+
+        for (int partition : partitions)
+            topicPartitions.add(new TopicPartition(topicName, partition));
+
+        return topicPartitions;
+    }
+
+    private SortedSet<TopicIdPartition> topicIdPartitions(Uuid topicId, String topicName, int... partitions) {
+        SortedSet<TopicIdPartition> topicIdPartitions = new TreeSet<>(new Utils.TopicIdPartitionComparator());
+
+        for (int partition : partitions)
+            topicIdPartitions.add(new TopicIdPartition(topicId, new TopicPartition(topicName, partition)));
+
+        return topicIdPartitions;
+    }
+
+    private void performCallback(MembershipManagerImpl membershipManager,
+                                 ConsumerRebalanceListenerInvoker invoker,
+                                 ConsumerRebalanceListenerMethodName expectedMethodName,
+                                 SortedSet<TopicPartition> expectedPartitions) {
+        // We expect only our enqueued event in the background queue.
+        assertEquals(1, backgroundEventQueue.size());
+        assertNotNull(backgroundEventQueue.peek());
+        assertInstanceOf(ConsumerRebalanceListenerCallbackNeededEvent.class, backgroundEventQueue.peek());
+        ConsumerRebalanceListenerCallbackNeededEvent neededEvent = (ConsumerRebalanceListenerCallbackNeededEvent) backgroundEventQueue.poll();
+        assertNotNull(neededEvent);
+        assertEquals(expectedMethodName, neededEvent.methodName());
+        assertEquals(expectedPartitions, neededEvent.partitions());
+
+        ConsumerRebalanceListenerCallbackCompletedEvent invokedEvent = invokeRebalanceCallbacks(
+                invoker,
+                neededEvent.methodName(),
+                neededEvent.partitions(),
+                neededEvent.future()
+        );
```

Comment:
I am sorry but I still don't get why we really need to call the invoker here. It seems simpler to just create the event:

```
        return new ConsumerRebalanceListenerCallbackCompletedEvent(
            methodName,
            future,
            error
        );
```

We would need to pass `error` to `performCallback`. If we do this, we could remove all the listener's related bits from all the tests and `CounterConsumerRebalanceListener`. I may be missing something...

Reply:
Sorry for adding to the confusion 😞 

Yes, I could just invoke the method on the `ConsumerRebalanceListener` instance directly, but by structuring the test this way, it exercises the following production code to help ensure correctness:

* `AsyncKafkaConsumer.invokeRebalanceCallbacks()` tests the correctness of the relevant `ConsumerRebalanceListenerMethodName`, error, and event handling logic
* `ConsumerRebalancerListenerInvoker.invokePartitions*` tests the correctness of the relevant `ConsumerRebalancerListener`, `SubscriptionState`, `Optional`, metrics, and log handling logic

Reply:
I see... I am definitely not a fan of doing this because a unit test is supposed to test the logic of the `unit`.  All the other ones that you mentioned should have their own unit tests and we should have integration tests for validating the end-to-end behaviour.